File: python/python-sdk/kestrapy/api/auditlogs_api.py
from typing import Any, Dict, List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.query_filter import QueryFilter


class AuditLogsApi(BaseApi):
    """Audit logs (`/api/v1/{tenant}/auditlogs` and the instance-owner
    cross-tenant variants under `/api/v1/auditlogs`).

    The controller is gated by the ``AUDITLOG`` resource plus a licence feature,
    so routes answer 403 otherwise.
    """

    def search_audit_logs(
        self,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> Dict[str, Any]:
        """A page of a tenant's audit log. GET /api/v1/{tenant}/auditlogs/search."""
        path = self._tenant_path(tenant, "auditlogs", "search")
        params = list(self._build_query_params(page=page, size=size).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._raw_json_request("GET", path, params=params)

    def search_all_audit_logs(
        self,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> Dict[str, Any]:
        """A page of the audit log across every tenant (instance-owner only).
        GET /api/v1/auditlogs/search."""
        path = self._superadmin_path("auditlogs", "search")
        params = list(self._build_query_params(page=page, size=size).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._raw_json_request("GET", path, params=params)

    def export_audit_logs(
        self, tenant: str, filters: Optional[List[QueryFilter]] = None
    ) -> str:
        """Stream a tenant's audit log as CSV.
        GET /api/v1/{tenant}/auditlogs/export."""
        path = self._tenant_path(tenant, "auditlogs", "export")
        params = []
        self._append_filter_params(params, filters)
        return self._text_request("GET", path, params=params, accept=self.CSV)

    def export_all_audit_logs(
        self, filters: Optional[List[QueryFilter]] = None
    ) -> str:
        """Stream the cross-tenant audit log as CSV (instance-owner only).
        GET /api/v1/auditlogs/export."""
        path = self._superadmin_path("auditlogs", "export")
        params = []
        self._append_filter_params(params, filters)
        return self._text_request("GET", path, params=params, accept=self.CSV)

    def audit_log_diff(
        self, id: str, tenant: str, previous_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """The before/after diff captured by an audit-log entry.
        GET /api/v1/{tenant}/auditlogs/{id}/diff."""
        path = self._tenant_path(tenant, "auditlogs", id, "diff")
        params = self._build_query_params(previousId=previous_id)
        return self._raw_json_request("GET", path, params=params)

    def global_audit_log_diff(
        self, id: str, previous_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """The diff for a global (non-tenant) resource audit entry
        (instance-owner only). GET /api/v1/auditlogs/{id}/diff."""
        path = self._superadmin_path("auditlogs", id, "diff")
        params = self._build_query_params(previousId=previous_id)
        return self._raw_json_request("GET", path, params=params)

    def find_audit_log(
        self, tenant: str, body: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Find a single audit-log entry matching a resource/action/detail query.
        ``body`` is a FindRequest ({resource, type?, detail}).
        POST /api/v1/{tenant}/auditlogs/find."""
        path = self._tenant_path(tenant, "auditlogs", "find")
        return self._raw_json_request("POST", path, body=body)

    def audit_log_history(
        self, detail_id: str, tenant: str
    ) -> List[Dict[str, Any]]:
        """The audit-log entries recorded for a single resource id.
        GET /api/v1/{tenant}/auditlogs/history/{detailId}."""
        path = self._tenant_path(tenant, "auditlogs", "history", detail_id)
        return self._raw_json_request("GET", path)

File: go-sdk/kestra_api_client/auditlogs_api.go
package kestra_api_client

import "context"

// AuditLogsAPI covers /api/v1/{tenant}/auditlogs and the instance-owner
// cross-tenant variants under /api/v1/auditlogs. The controller is gated by the
// AUDITLOG resource plus a license feature, so routes answer 403 otherwise.
type AuditLogsAPI struct {
	baseAPI
}

// SearchAuditLogs returns a page of a tenant's audit log. Backs GET
// /api/v1/{tenant}/auditlogs/search.
func (a *AuditLogsAPI) SearchAuditLogs(ctx context.Context, tenant string, page, size *int, sort []string, filters []SearchFilter) (map[string]interface{}, error) {
	params := buildQueryParams("page", page, "size", size)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "auditlogs", "search"), nil, params)
}

// SearchAllAuditLogs returns a page of the audit log across every tenant
// (instance-owner only). Backs GET /api/v1/auditlogs/search.
func (a *AuditLogsAPI) SearchAllAuditLogs(ctx context.Context, page, size *int, sort []string, filters []SearchFilter) (map[string]interface{}, error) {
	params := buildQueryParams("page", page, "size", size)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", superadminPath("auditlogs", "search"), nil, params)
}

// ExportAuditLogs streams a tenant's audit log as CSV. Backs GET
// /api/v1/{tenant}/auditlogs/export.
func (a *AuditLogsAPI) ExportAuditLogs(ctx context.Context, tenant string, filters []SearchFilter) (string, error) {
	params := buildQueryParams()
	appendFilterParams(params, filters)
	return a.doText(ctx, "GET", tenantPath(tenant, "auditlogs", "export"), params, contentCSV)
}

// ExportAllAuditLogs streams the cross-tenant audit log as CSV (instance-owner
// only). Backs GET /api/v1/auditlogs/export.
func (a *AuditLogsAPI) ExportAllAuditLogs(ctx context.Context, filters []SearchFilter) (string, error) {
	params := buildQueryParams()
	appendFilterParams(params, filters)
	return a.doText(ctx, "GET", superadminPath("auditlogs", "export"), params, contentCSV)
}

// AuditLogDiff returns the before/after diff captured by an audit-log entry.
// `previousId` is optional. Backs GET /api/v1/{tenant}/auditlogs/{id}/diff.
func (a *AuditLogsAPI) AuditLogDiff(ctx context.Context, id, tenant string, previousId *string) (map[string]interface{}, error) {
	params := buildQueryParams("previousId", previousId)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "auditlogs", id, "diff"), nil, params)
}

// GlobalAuditLogDiff returns the diff for a global (non-tenant) resource audit
// entry (instance-owner only). Backs GET /api/v1/auditlogs/{id}/diff.
func (a *AuditLogsAPI) GlobalAuditLogDiff(ctx context.Context, id string, previousId *string) (map[string]interface{}, error) {
	params := buildQueryParams("previousId", previousId)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", superadminPath("auditlogs", id, "diff"), nil, params)
}

// FindAuditLog finds a single audit-log entry matching a resource/action/detail
// query. `body` is a FindRequest ({resource, type?, detail}). Backs POST
// /api/v1/{tenant}/auditlogs/find.
func (a *AuditLogsAPI) FindAuditLog(ctx context.Context, tenant string, body interface{}) (map[string]interface{}, error) {
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "POST", tenantPath(tenant, "auditlogs", "find"), body, nil)
}

// AuditLogHistory lists the audit-log entries recorded for a single resource id.
// Backs GET /api/v1/{tenant}/auditlogs/history/{detailId}.
func (a *AuditLogsAPI) AuditLogHistory(ctx context.Context, detailId, tenant string) ([]map[string]interface{}, error) {
	return doJSON[[]map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "auditlogs", "history", detailId), nil, nil)
}
